src/ops/attention.cc

```diff
@@ -269,6 +263,12 @@ MultiHeadAttention::MultiHeadAttention(FFModel &model,
       qoSeqLength(_query->dims[1].size), kvSeqLength(_key->dims[1].size)
 // bias_initializer(_bias_initializer)
 {
+  // assert shape is valid
+  assert (_query->check_valid());
+  assert (_key->check_valid());
+  assert (_value->check_valid());
```

Comment:
Also, why does `check_valid` both return a `bool` and also do assertions in itself? Could you add some doxygen docs explaining that function, and simplify its behavior if possible?

Reply:
Yeah, I'm not actually sure why this was the case. I think we should replace the assertions in `check_valid()` with `return false` and keep the assertions in the operator implementations. `is_valid()` on `ParallelTensorShape` does this as well.

Reply:
Sounds good, just make sure you add assertions to any places where `check_valid` is currently called without a corresponding `assert`.